Adding new classes that create material property for strain and stres…

vanwdani opened the pull request:
…s within TM Master Action

Refs #14736
@bwspenc

vanwdani committed:
Adding new classes that create material property for strain and stress within TM Master Action
Ref #14736

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/466155/) on 5513f0c wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/14740/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/14740/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 48fe948cda1433e4138dc2d69360d91d0b362af5`

bwspenc commented:
You're going to need to add Markdown files to document these classes (the tests will fail on that soon!)

vanwdani committed:
Added documentation as well as util doc folder

vanwdani committed:
Resolved AD problem, removed un-necessary md files
Modified allowable strains for rspherical coordinates
Renamed classes to reflect outcome better
Modified RankTwoScalarTools enum calls
Review changes on names and calls to RankTwoScalarTools
AD class creation
Refs #14740

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/467290/) on b34ff0d wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/14740/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/14740/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 421fe11d8bf871a760ee908e5f229ec817612d38`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/467315/) on 768a66a wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/14740/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/14740/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 421fe11d8bf871a760ee908e5f229ec817612d38`

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/467810/) on 538a518 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/14740/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/469077/) on 6986e62 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/14740/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/14740/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 421fe11d8bf871a760ee908e5f229ec817612d38`

moosebuild commented:
Job [Modules debug](https://civet.inl.gov/job/469106/) on 46f6a0b : invalidated by @vanwdani

Hang up

moosebuild commented:
Job [Test debug](https://civet.inl.gov/job/469107/) on 46f6a0b : invalidated by @vanwdani

Hang up

moosebuild commented:
All [jobs](https://civet.inl.gov/event/51412/) on 46f6a0b : invalidated by @vanwdani

Hang up

moosebuild commented:
Job [Conda App tests](https://civet.inl.gov/job/469114/) on 46f6a0b : invalidated by @vanwdani

unrelated errors

moosebuild commented:
All [jobs](https://civet.inl.gov/event/51547/) on e5d98e9 : invalidated by @vanwdani

shutdown over weekend

moosebuild commented:
Job [Conda App tests](https://civet.inl.gov/job/472015/) on 42e10cd : invalidated by @vanwdani

Unrelated test failure

moosebuild commented:
All [jobs](https://civet.inl.gov/event/51823/) on b73dc63 : invalidated by @vanwdani

moosebuild commented:
Job [Python Tests](https://civet.inl.gov/job/474665/) on b73dc63 : invalidated by @vanwdani

moosebuild commented:
Job [Conda App tests](https://civet.inl.gov/job/474686/) on b73dc63 : invalidated by @vanwdani

moosebuild commented:
Job [Python Tests](https://civet.inl.gov/job/474665/) on b73dc63 : invalidated by @vanwdani

moosebuild commented:
Job [Conda App tests](https://civet.inl.gov/job/474686/) on b73dc63 : invalidated by @vanwdani

bwspenc commented on modules/tensor_mechanics/doc/content/source/utils/RankTwoScalarTools.md:
```diff
@@ -0,0 +1,247 @@
+
+## Description
+
+Calls to this class calculates the following strain/stress quantities.
```
I would say this instead:

This is a utility class that provides a set of functions to compute scalar quantities for stress or strain tensors. This  class is not directly invoked in the input file, but is called by several other classes that are, including <provide list of links to them>.

bwspenc reviewed (changes requested):
I think we need to work on the basic set of classes a little more.  I'll come by to discuss.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/494033/) on f5e1610 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/14740/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/14740/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 55a2368dca45913ae0385c4d5fcff68318d3e4ef`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/494057/) on d8204fe wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/14740/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/14740/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 55a2368dca45913ae0385c4d5fcff68318d3e4ef`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/494081/) on a45c7e3 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/14740/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/14740/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 55a2368dca45913ae0385c4d5fcff68318d3e4ef`

bwspenc commented on modules/tensor_mechanics/src/materials/MaterialRankTwoDirectionScalar.C:
```diff
@@ -0,0 +1,63 @@
+//* This file is part of the MOOSE framework
```
It doesn't look like MaterialRankTwoDirectionScalar and MaterialRankTwoScalar are different in any substantial way.  What I had in mind was to have a set of models, each which only exposed the parameters applicable to it. High level things to change:
1) Drop 'Material' from the beginning of the name of all of these
2) If you just want to mimic what we currently do for the AuxKernels, you could have just two new classes: `RankTwoTensorScalar` and `RankTwoTensorComponent`. You could accomplish that by renaming `MaterialRankTwoTensorComponent` -> `RankTwoTensorComponent` and `MaterialRankTwoDirectionScalar`->`RankTwoTensorScalar` and deleting `MaterialRankTwoScalar`.
3) You could stop at 2) and have something like what we currently have, but we can do a little better than that by splitting the (renamed) `RankTwoTensorScalar` into a few separate classes that only have required parameters:
a) A class that computes the scalar reductions that don't require `point1`, `point2`, or `direction`. Don't provide those as parameters. You could use `RankTwoTensorScalar` as the name for this one. 
b) A class that computes the stress in a specified direction, and requires `direction`. You could call it `RankTwoScalarDirectionalComponent`
c) A class that computes the radial, hoop, or axial stress and requires `point1` and `point2`. You could call it `RankTwoCylindricalComponent`

dschwen commented on modules/tensor_mechanics/include/materials/MaterialRankTwoDirectionScalar.h:
```diff
@@ -0,0 +1,48 @@
+#include "Material.h"
+#include "RankTwoTensor.h"
+
+class MaterialRankTwoDirectionScalar;
+
+template <>
+InputParameters validParams<MaterialRankTwoDirectionScalar>();
```
Remove these lines

dschwen commented on modules/tensor_mechanics/include/materials/MaterialRankTwoScalar.h:
```diff
@@ -0,0 +1,48 @@
+#include "Material.h"
+#include "RankTwoTensor.h"
+
+class MaterialRankTwoScalar;
+
+template <>
+InputParameters validParams<MaterialRankTwoScalar>();
```
Remove these lines

dschwen commented on modules/tensor_mechanics/include/materials/MaterialRankTwoTensorComponent.h:
```diff
@@ -0,0 +1,45 @@
+#include "Material.h"
+#include "RankTwoTensor.h"
+
+class MaterialRankTwoTensorComponent;
```
remove

dschwen commented on modules/tensor_mechanics/include/materials/MaterialRankTwoTensorComponent.h:
```diff
@@ -0,0 +1,45 @@
+
+template <>
+InputParameters validParams<MaterialRankTwoTensorComponent>();
+
```
remove the empty line

dschwen commented on modules/tensor_mechanics/src/actions/TensorMechanicsAction.C:
```diff
@@ -235,12 +237,14 @@ TensorMechanicsAction::act()
     std::string type;
 
+    //  actOutputMatProp();
+
```
remove commented out code

dschwen commented on modules/tensor_mechanics/src/materials/MaterialRankTwoDirectionScalar.C:
```diff
@@ -0,0 +1,63 @@
+
+registerMooseObject("TensorMechanicsApp", MaterialRankTwoDirectionScalar);
+
+defineLegacyParams(MaterialRankTwoDirectionScalar);
```
remove

dschwen commented on modules/tensor_mechanics/src/materials/MaterialRankTwoScalar.C:
```diff
@@ -0,0 +1,66 @@
+
+registerMooseObject("TensorMechanicsApp", MaterialRankTwoScalar);
+
+defineLegacyParams(MaterialRankTwoScalar);
```
remove

dschwen commented on modules/tensor_mechanics/src/materials/MaterialRankTwoTensorComponent.C:
```diff
@@ -0,0 +1,62 @@
+
+registerMooseObject("TensorMechanicsApp", MaterialRankTwoTensorComponent);
+
+defineLegacyParams(MaterialRankTwoTensorComponent);
```
remove

bwspenc commented:
I don't think these class names should start with "Material". Let's talk about it.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/499851/) on 1f5f6e3 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/14740/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/14740/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format d1bab0f6ae1c62ee15640d5115af75565366a98d`

bwspenc commented on modules/tensor_mechanics/doc/content/source/materials/RankTwoScalarDirectionalComponent.md:
```diff
@@ -0,0 +1,25 @@
+
+## Description
+
+Rank Two Scalar Directional Component is used within
```
Make changes similar to what I suggested for RankTwoCylindricalComponent.

bwspenc commented on modules/tensor_mechanics/doc/content/source/materials/RankTwoTensorComponent.md:
```diff
@@ -0,0 +1,31 @@
+
+## Description
+
+Rank Two Tensor Component is used within
```
Make changes similar to what I suggested for RankTwoCylindricalComponent.

bwspenc commented on modules/tensor_mechanics/doc/content/source/materials/RankTwoTensorScalar.md:
```diff
@@ -0,0 +1,22 @@
+
+## Description
+
+Rank Two Tensor Scalar is used within
```
Make changes similar to what I suggested for RankTwoCylindricalComponent.

bwspenc commented on modules/tensor_mechanics/doc/content/source/utils/RankTwoScalarTools.md:
```diff
@@ -0,0 +1,249 @@
+## Description
+
+This is a utility class that provides a set of functions to compute scalar
+quantities for stress or strain tensors. This class is not directly invoked in
```
'for stress or strain tensors' -> 'such as invariants and components in specified directions from rank-2 tensors such as stress or strain.'

dschwen replied:
Also `RankTwoScalarTools` is _not_ a class. It is a namespaced collection of helper functions.

bwspenc commented on modules/tensor_mechanics/doc/content/source/utils/RankTwoScalarTools.md:
```diff
@@ -0,0 +1,249 @@
+
+This is a utility class that provides a set of functions to compute scalar
+quantities for stress or strain tensors. This class is not directly invoked in
+the input file, but is called by several other classes that are, including:
```
I think you meant to list out the classes here.  That's a good idea.  They would include the 4 material models you are adding, as well as the 2 pre-existing AuxKernels.

After that, then introduce the list of scalar quantities that can be computed.

bwspenc commented on modules/tensor_mechanics/doc/content/source/utils/RankTwoScalarTools.md:
```diff
@@ -0,0 +1,249 @@
+# Rank Two Scalar Tools
```
This material looks really familiar. Is it copied from another location? Maybe the AuxKernels that use it? We should remove it from that other place and provide a link to this page from there if it is.

bwspenc commented on modules/tensor_mechanics/include/actions/TensorMechanicsAction.h:
```diff
@@ -93,6 +92,9 @@ class TensorMechanicsAction : public TensorMechanicsActionBase
+  /// output materials to generate scalar stress/strain tensor quantities
   std::vector<std::string> _generate_output;
+
+  const Point _point1;
```
Add doxygen description of variables

bwspenc commented on modules/tensor_mechanics/include/materials/RankTwoCylindricalComponent.h:
```diff
@@ -0,0 +1,43 @@
+#include "RankTwoTensor.h"
+
+/**
+ * MaterialRankTwoCylindricalScalar uses the namespace RankTwoScalarTools to compute scalar
```
This description isn't right.

bwspenc commented on modules/tensor_mechanics/include/materials/RankTwoCylindricalComponent.h:
```diff
@@ -0,0 +1,43 @@
+  virtual void initQpStatefulProperties() override;
+  virtual void computeQpProperties() override;
+
+  const MaterialProperty<RankTwoTensor> & _tensor;
```
Add doxygen descriptions of variables. Use the one-liner form (`///`) for all of them for consistency.

bwspenc commented on modules/tensor_mechanics/include/materials/RankTwoScalarDirectionalComponent.h:
```diff
@@ -0,0 +1,41 @@
+  virtual void initQpStatefulProperties() override;
+  virtual void computeQpProperties() override;
+
+  const MaterialProperty<RankTwoTensor> & _tensor;
```
doxygen

bwspenc commented on modules/tensor_mechanics/include/materials/RankTwoTensorComponent.h:
```diff
@@ -0,0 +1,39 @@
+  virtual void computeQpProperties() override;
+
+private:
+  const MaterialProperty<RankTwoTensor> & _tensor;
```
doxygen

bwspenc commented on modules/tensor_mechanics/include/materials/RankTwoTensorScalar.h:
```diff
@@ -0,0 +1,39 @@
+  virtual void initQpStatefulProperties() override;
+  virtual void computeQpProperties() override;
+
+  const MaterialProperty<RankTwoTensor> & _tensor;
```
doxygen

bwspenc commented on modules/tensor_mechanics/include/utils/RankTwoScalarTools.h:
```diff
@@ -467,4 +470,83 @@ getQuantity(const RankTwoTensorTempl<T> & tensor,
+  switch (scalar_type)
+  {
+    case 0:
+      return maxPrincipal(tensor, direction);
```
I think the principal stresses don't belong in this function. The principal stresses are considered invariants of the stress tensor because you don't supply a direction -- it comes out of the calculation. 

This method should have direction be `const`, while it would not be const for the function that computes principal stresses.

bwspenc commented on modules/tensor_mechanics/src/actions/TensorMechanicsAction.C:
```diff
@@ -76,7 +76,9 @@ TensorMechanicsAction::TensorMechanicsAction(const InputParameters & params)
         getParam<MooseEnum>("out_of_plane_direction").getEnum<OutOfPlaneDirection>()),
-    _base_name(isParamValid("base_name") ? getParam<std::string>("base_name") + "_" : "")
+    _base_name(isParamValid("base_name") ? getParam<std::string>("base_name") + "_" : ""),
+    _point1(isParamValid("point1") ? getParam<Point>("point1") : Point(0, 0, 0)),
```
I don't think we should have default values for the points that define the axis of rotation. If you want a cylindrical stress component, you would need to specify these. Also, this is a good opportunity to change the names of these parameters to something more descriptive of what they are, because `point1` and `point2` don't mean much in the context of the full set of parameters used in this Action.

Maybe we could specify `cylindrical_axis_point1` and `cylindrical_axis_point2`. Alternatively, we could just have the user specify `cylindrical_axis`, and provide `x`, `y`, or `z` as an enum in this context. It's less general, but simpler. These models are almost always set up with the axis of rotation being one of those axes. Those could be translated to two points to pass in to the Material that uses them. If the user has the more general case, they could set up the Materials themselves. Thoughts, @dschwen, @sapitts?

bwspenc commented on modules/tensor_mechanics/src/actions/TensorMechanicsAction.C:
```diff
@@ -468,47 +472,124 @@ TensorMechanicsAction::actOutputGeneration()
     {
       std::string type = "";
```
`type` isn't used anymore, is it?

bwspenc commented on modules/tensor_mechanics/src/actions/TensorMechanicsAction.C:
```diff
@@ -468,47 +472,124 @@ TensorMechanicsAction::actOutputGeneration()
+    ad_append = "<RESIDUAL>";
+  }
+  //
+  // Create materials which allow the use of material properties for quatities
```
Drop this comment -- it doesn't add much.

dschwen replied:
it adds a typo...

bwspenc commented on modules/tensor_mechanics/src/actions/TensorMechanicsAction.C:
```diff
@@ -468,47 +472,124 @@ TensorMechanicsAction::actOutputGeneration()
+          }
+      //
+      // This material property is already created by other materials such as
+      // ZryCreepLimbackHoppeUpdate.
```
That's a BISON model name.  Don't reference here.  You could just say 'by creep or plasticity models'.

bwspenc commented on modules/tensor_mechanics/src/materials/RankTwoCylindricalComponent.C:
```diff
@@ -0,0 +1,60 @@
+  params.addParam<MooseEnum>(
+      "scalar_type", RankTwoScalarTools::cylindricalOptions(), "Type of scalar output");
+  params.addParam<Point>(
+      "point1", "Start point for axis used to calculate some Direction material tensor quantities");
```
I'd name these `cylindrical_axis_point1` and `cylindrical_axis_point2`, and change the descriptions accordingly.

bwspenc commented on modules/tensor_mechanics/src/materials/RankTwoCylindricalComponent.C:
```diff
@@ -0,0 +1,60 @@
+RankTwoCylindricalComponent::validParams()
+{
+  InputParameters params = Material::validParams();
+  params.addClassDescription("Compute a cylindrical stress/strain property of a RankTwoTensor");
```
Compute components of a rank-2 tensor in a cylindrical coordinate system.

bwspenc commented on modules/tensor_mechanics/src/materials/RankTwoCylindricalComponent.C:
```diff
@@ -0,0 +1,60 @@
+  params.addClassDescription("Compute a cylindrical stress/strain property of a RankTwoTensor");
+  params.addRequiredParam<MaterialPropertyName>("rank_two_tensor",
+                                                "The rank two material tensor name");
+  params.addParam<std::string>("calculation_name",
```
Rename `calculation_name` to `property_name` for consistency with other models. Also, the description should just be something like: 'Name of the material property computed by this model'

bwspenc commented on modules/tensor_mechanics/src/materials/RankTwoCylindricalComponent.C:
```diff
@@ -0,0 +1,60 @@
+RankTwoCylindricalComponent::RankTwoCylindricalComponent(const InputParameters & parameters)
+  : Material(parameters),
+    _tensor(getMaterialProperty<RankTwoTensor>("rank_two_tensor")),
+    _calculation_name(isParamValid("calculation_name") ? getParam<std::string>("calculation_name")
```
`_calculation_name` can't be an empty string.  I'd say that `property_name` should probably be required. Either that, or you could automatically come up with something like `hoop_stress` based on a combination of the component they requested and the tensor name, but I don't think that's worth the trouble.

bwspenc commented on modules/tensor_mechanics/src/materials/RankTwoCylindricalComponent.C:
```diff
@@ -0,0 +1,60 @@
+    _tensor(getMaterialProperty<RankTwoTensor>("rank_two_tensor")),
+    _calculation_name(isParamValid("calculation_name") ? getParam<std::string>("calculation_name")
+                                                       : ""),
+    _calculation(declareProperty<Real>(_calculation_name)),
```
Rename `_calculation` to `_property`

bwspenc commented on modules/tensor_mechanics/src/materials/RankTwoCylindricalComponent.C:
```diff
@@ -0,0 +1,60 @@
+    _scalar_type(getParam<MooseEnum>("scalar_type")),
+    _point1(isParamValid("point1") ? getParam<Point>("point1") : Point(0, 0, 0)),
+    _point2(isParamValid("point2") ? getParam<Point>("point2") : Point(0, 1, 0)),
+    _direction(Point(0, 0, 0))
```
You shouldn't need a `_direction` variable.

bwspenc replied:
Based on our discussion, I guess you need this. I think that it's a little confusing for it to be a member variable because it makes it look like it's important. I think I would just create a dummy Point object just before the call to `getCylindricalComponent()`. You don't need to initialize it or anything, just do something like:
```
Point dummy_direction;

bwspenc commented on modules/tensor_mechanics/src/materials/RankTwoCylindricalComponent.C:
```diff
@@ -0,0 +1,60 @@
+  unsigned int qp = _qp;
+
+  _calculation[_qp] = RankTwoScalarTools::getCylindricalComponent(
+      _tensor[qp], _scalar_type, _point1, _point2, _q_point[qp], _direction);
```
This function shouldn't require `_direction`.

bwspenc commented on modules/tensor_mechanics/src/materials/RankTwoCylindricalComponent.C:
```diff
@@ -0,0 +1,60 @@
+                               "multiple mechanics material systems on the same "
+                               "block, i.e. for multiple phases");
+  params.addParam<MooseEnum>(
+      "scalar_type", RankTwoScalarTools::cylindricalOptions(), "Type of scalar output");
```
`scalar_type` should be renamed to something more descriptive like `cylindrical_component` or just `component`

bwspenc commented on modules/tensor_mechanics/src/materials/RankTwoTensorComponent.C:
```diff
@@ -0,0 +1,60 @@
+  params.addClassDescription("Access a component of a RankTwoTensor");
+  params.addRequiredParam<MaterialPropertyName>("rank_two_tensor",
+                                                "The rank two material tensor name");
+  params.addParam<std::string>("calculation_name",
```
Change `calculation_name` to `property_name` and other related changes consistent with what I suggested for the cylindrical class.

bwspenc commented on modules/tensor_mechanics/src/materials/RankTwoTensorScalar.C:
```diff
@@ -0,0 +1,52 @@
+  params.addClassDescription("Compute a scalar property of a RankTwoTensor");
+  params.addRequiredParam<MaterialPropertyName>("rank_two_tensor",
+                                                "The rank two material tensor name");
+  params.addParam<std::string>("calculation_name",
```
Same deal -- change to `property_name`

bwspenc commented on modules/tensor_mechanics/src/materials/RankTwoTensorScalar.C:
```diff
@@ -0,0 +1,52 @@
+                               "multiple mechanics material systems on the same "
+                               "block, i.e. for multiple phases");
+  params.addParam<MooseEnum>(
+      "scalar_type", RankTwoScalarTools::tensorScalarOptions(), "Type of scalar output");
```
I would suggest changing `scalar_type` to `invariant`

bwspenc commented on modules/tensor_mechanics/src/utils/RankTwoScalarTools.C:
```diff
@@ -21,6 +23,26 @@ scalarOptions()
+MooseEnum
+scalarDirectionalOptions()
+{
+  return MooseEnum("MaxPrincipal MidPrincipal MinPrincipal Direction");
```
I'd re-name this to `principalComponentOptions()` and get rid of `Direction`.

dschwen commented on modules/tensor_mechanics/doc/content/source/materials/RankTwoTensorComponent.md:
```diff
@@ -0,0 +1,31 @@
+`index_i` and the `index_j` for the single tensor component to save into an
+MaterialProperty.  [eq:rank2tensor_component_indices] shows the index values
+for each Rank-2 tensor component.
+\begin{equation}
```
the 
```
\begin{equation}
\end{equation}
```

syntax is going away. Replace with

```
!equation
\label{eq:rank2tensor_component_indices}
\sigma_{ij} \implies \begin{bmatrix}
                      \sigma_{00} & \sigma_{01} & \sigma_{02} \\
                      \sigma_{10} & \sigma_{11} & \sigma_{12} \\
                      \sigma_{20} & \sigma_{21} & \sigma_{22}
                      \end{bmatrix}

```

dschwen commented on modules/tensor_mechanics/doc/content/source/utils/RankTwoScalarTools.md:
```diff
@@ -0,0 +1,249 @@
+# Rank Two Scalar Tools
+
+
```
two empty lines?

sapitts commented on modules/tensor_mechanics/doc/content/source/materials/RankTwoCylindricalComponent.md:
```diff
@@ -0,0 +1,25 @@
+#Rank Two Sclara Cylindrical Component
```
Please do retain the spaces among the words, particularly for these longer class names. The text wrapping for longer class names does not work when spaces are not present in the class name and the class name overruns into the right side bar

bwspenc replied:
@sapitts I think we should revisit this.  The search function doesn't work for classes that have spaces.

bwspenc replied:
That topic is tangential to this PR, though.  We can leave the spaces for now for consistency with other classes, but I'd like to have a discussion about this.

sapitts replied:
Okay. I knew that interjecting my opinion here wasn't going to change anything. I had hoped to raise the profile of the longer page titles over running the right side bar. I agree that it is a tangential issue to this PR

@vanwdani looks like you will need a space between the pound sign and the first letter of the class name to get this to render properly on the documentation website.

dschwen replied:
To chime in with my two cents, I'm not fond of the spaces, or titles that are not the class name for that matter (unless its on auxiliary doc pages of course).

dschwen replied:
This was marked as resolved, but still says "Sclara". That is wrong no matter what side you're on regarding the spaces.

vanwdani replied:
I mark resolved on my end to keep track of the changes I have made before pushing.  The typo has been removed, but the spaces have remained as of now

dschwen commented on modules/tensor_mechanics/doc/content/source/utils/RankTwoScalarTools.md:
```diff
@@ -0,0 +1,249 @@
+
+## Axial Stress
+
+The scalar type `AxialStress` calculates the scalar value of a Rank-2 tensor,
```
The scalar type does not compute anything. These enum values are only relevant for the calling class. In this file you should describe the actual functions in this file.

dschwen commented on modules/tensor_mechanics/doc/content/syntax/Modules/TensorMechanics/Master/index.md:
```diff
@@ -25,7 +25,7 @@ The Tensor Mechanics `Master` Action is used to construct the kernels, displacem
 |   | [ComputeSmallStrain](/ComputeSmallStrain.md) or [ComputePlaneSmallStrain](/ComputePlaneSmallStrain.md) or [ComputeAxisymmetric1DSmallStrain](/ComputeAxisymmetric1DSmallStrain.md) or [ComputeAxisymmetricRZSmallStrain](/ComputeAxisymmetricRZSmallStrain.md) |   |
 |   | [ComputeIncrementalSmallStrain](/ComputeIncrementalSmallStrain.md) or [ComputePlaneIncrementalStrain](/ComputePlaneIncrementalStrain.md) or [ComputeAxisymmetric1DIncrementalStrain](/ComputeAxisymmetric1DIncrementalStrain.md) or [ComputeAxisymmetricRZIncrementalStrain](/ComputeAxisymmetricRZIncrementalStrain.md) | `incremental` : boolean for using a incremental strain formulation |
-| Add AuxVariables and AuxKernels for various tensor component and quantity outputs | [AuxVariables](/AuxVariables/index.md) and [RankTwoAux](/RankTwoAux.md) or [RankTwoScalarAux](/RankTwoScalarAux.md) or [RankFourAux](/RankFourAux.md) | `generate_output`: a string of the quantities to add |
+| Add AuxVariables and AuxKernels for various tensor component and quantity outputs | Material Properties as well as [AuxVariables](/AuxVariables/index.md) and [RankTwoAux](/RankTwoAux.md) or [RankTwoScalarAux](/RankTwoScalarAux.md) or [RankFourAux](/RankFourAux.md) | `generate_output`: a string of the quantities to add |
```
separate line for the material properties output option

dschwen commented on modules/tensor_mechanics/include/actions/TensorMechanicsActionBase.h:
```diff
@@ -20,12 +20,18 @@ class TensorMechanicsActionBase : public Action
 
   static MultiMooseEnum outputPropertiesType();
 
+  static MultiMooseEnum unrealisiticRSphericalCalls();
```
What is this? It doesn't seem to get implemented and the name doesn't tell me what this is supposed to do.

dschwen commented on modules/tensor_mechanics/include/actions/TensorMechanicsActionBase.h:
```diff
@@ -20,12 +20,18 @@ class TensorMechanicsActionBase : public Action
   static const std::vector<char> _component_table;
   static const std::map<std::string, std::pair<std::string, std::vector<std::string>>>
-      _ranktwoscalaraux_table;
+      _ranktwo_tensor_scalar_table;
```
either `_rank_two_tensor_..` or `_ranktwotensor_...`

dschwen commented on modules/tensor_mechanics/include/materials/RankTwoCylindricalComponent.h:
```diff
@@ -0,0 +1,43 @@
+  const std::string _calculation_name;
+
+  MaterialProperty<Real> & _calculation;
+  /**
```
space above

dschwen commented on modules/tensor_mechanics/include/materials/RankTwoScalarDirectionalComponent.h:
```diff
@@ -0,0 +1,41 @@
+ * RankTwoScalarDirectionalComponent uses the namespace RankTwoScalarTools to compute scalar
+ * values from Rank-2 tensors.
+ */
+class RankTwoScalarDirectionalComponent : public Material
```
I like the favorite set

dschwen commented on modules/tensor_mechanics/include/materials/RankTwoScalarDirectionalComponent.h:
```diff
@@ -0,0 +1,41 @@
+  const std::string _calculation_name;
+
+  MaterialProperty<Real> & _calculation;
+  /**
```
space above

dschwen commented on modules/tensor_mechanics/include/materials/RankTwoTensorScalar.h:
```diff
@@ -0,0 +1,39 @@
+  const std::string _calculation_name;
+
+  MaterialProperty<Real> & _calculation;
+  /**
```
space above

dschwen commented on modules/tensor_mechanics/include/utils/RankTwoScalarTools.h:
```diff
@@ -467,4 +470,83 @@ getQuantity(const RankTwoTensorTempl<T> & tensor,
+template <typename T>
+T
+getCylindricalComponent(const RankTwoTensorTempl<T> & tensor,
+                        const MooseEnum & scalar_type,
```
My problem with this is that `MooseEnum`s are not strongly typed. Nothing would prevent a user to pass in a _scalar directional component_ `MooseEnum` into that function. As you are only querying index numbers, this will give a result. I would suggest using `enum class`es here. My preferred pattern is to never `getParam` a `MooseEnum` into a `MosseEnum` member variable, but rather use the `getParam("moose_enum_name").getEnum<EnumClassType>()` and thus have the parameter value in a strongly typed member variable. Then the parameter of this function would be a _specific_ enum class.

dschwen commented on modules/tensor_mechanics/src/actions/TensorMechanicsAction.C:
```diff
@@ -468,47 +472,124 @@ TensorMechanicsAction::actOutputGeneration()
-        for (const auto & t : r2sa.second.second)
-          if (r2sa.first + '_' + t == out)
+      //
+      // RankTwoTensorScalar
```
what is that comment style? no

dschwen commented on modules/tensor_mechanics/src/materials/RankTwoScalarDirectionalComponent.C:
```diff
@@ -0,0 +1,56 @@
+void
+RankTwoScalarDirectionalComponent::computeQpProperties()
+{
+  unsigned int qp = _qp;
```
remove and use `_qp` everywhere

dschwen commented on modules/tensor_mechanics/src/materials/RankTwoTensorComponent.C:
```diff
@@ -0,0 +1,60 @@
+void
+RankTwoTensorComponent::computeQpProperties()
+{
+  unsigned int qp = _qp;
```
same

dschwen commented on modules/tensor_mechanics/src/materials/RankTwoTensorScalar.C:
```diff
@@ -0,0 +1,52 @@
+void
+RankTwoTensorScalar::computeQpProperties()
+{
+  unsigned int qp = _qp;
```
same

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/501776/) on ad6c85b wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/14740/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/14740/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format d1bab0f6ae1c62ee15640d5115af75565366a98d`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/501805/) on 96d3bbe wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/14740/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/14740/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format e9feae96fed5e4e4fb115a5f35f35fa492e673a4`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/502064/) on f2e76db wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/14740/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/14740/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format e9feae96fed5e4e4fb115a5f35f35fa492e673a4`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/503086/) on 80cef3a wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/14740/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/14740/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format e9feae96fed5e4e4fb115a5f35f35fa492e673a4`

vanwdani committed:
New AD work-around
Refs #14736

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/509902/) on f40c10f wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/14740/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/14740/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format ee2af0766e3984acb4b23e5528518633c0266691`

vanwdani committed:
Clang formatting
Refs #14736

bwspenc commented on modules/tensor_mechanics/doc/content/source/auxkernels/RankTwoScalarAux.md:
```diff
@@ -4,376 +4,10 @@
-s = \sqrt{\frac{3}{2} S_{ij} S_{ij}}
-\end{equation}
-where $S_{ij}$ is the deviatoric tensor of the Rank-2 tensor $T_{ij}$.
+This is a namespace that provides a set of functions to compute scalar
```
No -- this is the AuxKernel that allows those quantities to be computed using RankTwoScalarTools. This isn't a namespace. I wouldn't use the word `namespace` anywhere in the documentation.  That will just confuse people.

It's fine to just have a short summary like this here and link to RankTwoScalar tools, but those issues need to be straightened out.

bwspenc commented on modules/tensor_mechanics/doc/content/source/materials/ADRankTwoCartesianComponent.md:
```diff
@@ -0,0 +1,34 @@
+Cartesian coordinate system. This can be used regardless of the coordinate
+system used by the model.
+
+This ADMaterial model is used by
```
is used by -> is set up by

bwspenc commented on modules/tensor_mechanics/doc/content/source/materials/ADRankTwoCartesianComponent.md:
```diff
@@ -0,0 +1,34 @@
+
+This ADMaterial model is used by
+[TensorMechanics/Master](/Modules/TensorMechanics/Master/index.md)
+automatically, but can also be called directly by the user. This class provides
```
automatically when stress components are requested in the `generate_output` parameter, but...

bwspenc replied:
called directly -> set up directly

bwspenc commented on modules/tensor_mechanics/doc/content/source/materials/ADRankTwoCartesianComponent.md:
```diff
@@ -0,0 +1,34 @@
+[TensorMechanics/Master](/Modules/TensorMechanics/Master/index.md)
+automatically, but can also be called directly by the user. This class provides
+methods to calculate several different scalar stress ($\boldsymbol{\sigma}$) and
+strain ($\boldsymbol{\epsilon}$)quantities for a Rank-2 tensor, as described in
```
Delete this sentence. This just grabs components, and doesn't even use RankTwoScalarTools.

vanwdani replied:
RankTwoScalarTools::component is the function used to determine the components.

bwspenc commented on modules/tensor_mechanics/doc/content/source/materials/ADRankTwoCylindricalComponent.md:
```diff
@@ -0,0 +1,27 @@
+cylindrical coordinate system. This can be used regardless of the coordinate
+system used by the model.
+
+This ADMaterial model is used by
```
Make the same minor tweaks (is used by, called directly, etc) that I suggested for the model above.

bwspenc commented on modules/tensor_mechanics/doc/content/source/materials/ADRankTwoDirectionalComponent.md:
```diff
@@ -0,0 +1,28 @@
+This ADMaterial model is used by
+[TensorMechanics/Master](/Modules/TensorMechanics/Master/index.md)
+automatically, but can also be called directly by the user. This class
+calculates the scalar value of a Rank-2 tensor, $T$, in the direction selected
```
Make the same tweaks as for the other ones. 

Also, change 'calculates the scalar value' to 'calculates the component'

bwspenc commented on modules/tensor_mechanics/doc/content/source/materials/ADRankTwoInvariant.md:
```diff
@@ -0,0 +1,30 @@
+
+This ADMaterial model is used by
+[TensorMechanics/Master](/Modules/TensorMechanics/Master/index.md)
+automatically, but can also be called directly by the user. This class provides
```
Same minor tweaks here.

bwspenc commented on modules/tensor_mechanics/doc/content/source/materials/RankTwoCartesianComponent.md:
```diff
@@ -0,0 +1,34 @@
+
+This Material model is used by
+[TensorMechanics/Master](/Modules/TensorMechanics/Master/index.md)
+automatically, but can also be called directly by the user. This class provides
```
Same minor tweaks here, and in the remaining ones.

bwspenc commented on modules/tensor_mechanics/doc/content/source/utils/RankTwoScalarTools.md:
```diff
@@ -0,0 +1,254 @@
+
+## Description
+
+This is a namespace that provides a set of functions to compute scalar
```
This is a set of functions

bwspenc commented on modules/tensor_mechanics/doc/content/source/utils/RankTwoScalarTools.md:
```diff
@@ -0,0 +1,254 @@
+
+This is a namespace that provides a set of functions to compute scalar
+quantities such as invariants and components in specified directions from rank-2
+tensors such as stress or strain. This class is not directly invoked in the
```
These functions are not directly invoked

bwspenc commented on modules/tensor_mechanics/doc/content/source/utils/RankTwoScalarTools.md:
```diff
@@ -0,0 +1,254 @@
+This is a namespace that provides a set of functions to compute scalar
+quantities such as invariants and components in specified directions from rank-2
+tensors such as stress or strain. This class is not directly invoked in the
+input file, but is called by several other classes such as:
```
but are called

bwspenc commented on modules/tensor_mechanics/doc/content/source/utils/RankTwoScalarTools.md:
```diff
@@ -0,0 +1,254 @@
+RankTwoTensorComponent, RankTwoCartesian, RankTwoInvariant, RankTwoScalarAux,
+and RankTwoAux.
+
+The scalar quantities that are computed from these classes:
```
The scalar quantities that can be computed include:

bwspenc commented on modules/tensor_mechanics/include/actions/TensorMechanicsAction.h:
```diff
@@ -93,6 +96,11 @@ class TensorMechanicsAction : public TensorMechanicsActionBase
+  /// output materials to generate scalar stress/strain tensor quantities
   std::vector<std::string> _generate_output;
+
+  /// points used to determine axis of rotation for cyclindrical stress/strain quantities
```
spelling error on cylindrical. Also, use the doxygen grouping.

bwspenc commented on modules/tensor_mechanics/include/materials/ADRankTwoCartesianComponent.h:
```diff
@@ -0,0 +1,44 @@
+
+/**
+ * ADRankTwoCartesianComponent is designed to take the data in the RankTwoTensor material
+ * property, for example Vonmises stress or strain, and output the value for the
```
No. It oututs a selected component. No vonMises.

bwspenc commented on modules/tensor_mechanics/include/materials/ADRankTwoCartesianComponent.h:
```diff
@@ -0,0 +1,44 @@
+  /// Stress/strain value returned from calculation
+  ADMaterialProperty<Real> & _property;
+
+  /// Tensor components
```
Use doxygen grouping

bwspenc commented on modules/tensor_mechanics/include/materials/ADRankTwoCylindricalComponent.h:
```diff
@@ -0,0 +1,50 @@
+#include "RankTwoTensor.h"
+
+/**
+ * MaterialRankTwoCylindricalComponent uses the namespace RankTwoScalarTools to compute scalar
```
Wrong class name, and this isn't what it does.

bwspenc commented on modules/tensor_mechanics/include/materials/ADRankTwoCylindricalComponent.h:
```diff
@@ -0,0 +1,50 @@
+   */
+  MooseEnum _cylindrical_component;
+
+  /// Points used to determine the axis of rotation
```
Doxygen grouping

bwspenc commented on modules/tensor_mechanics/include/materials/ADRankTwoCylindricalComponent.h:
```diff
@@ -0,0 +1,50 @@
+  /// Stress/strain value returned from calculation
+  ADMaterialProperty<Real> & _property;
+
+  /**
```
Use the single line `///` comments for describing variables, even if it takes a couple of lines.

bwspenc commented on modules/tensor_mechanics/include/materials/ADRankTwoDirectionalComponent.h:
```diff
@@ -0,0 +1,47 @@
+   * Determines the information to be extracted from the tensor by using the
+   * RankTwoScalarTools namespace
+   */
+  MooseEnum _invariant;
```
You don't even need this in this class, do you?  All it computes is the directional component.

bwspenc commented on modules/tensor_mechanics/include/materials/ADRankTwoDirectionalComponent.h:
```diff
@@ -0,0 +1,47 @@
+#include "RankTwoTensor.h"
+
+/**
+ * RankTwoDirectionalComponent uses the namespace RankTwoScalarTools to compute scalar
```
it computes the component of the tensor in a specified direction.

bwspenc commented on modules/tensor_mechanics/include/materials/ADRankTwoInvariant.h:
```diff
@@ -0,0 +1,46 @@
+#include "RankTwoTensor.h"
+
+/**
+ * RankTwoInvariant uses the namespace RankTwoScalarTools to compute scalar
```
It computes invariants of RankTwo Tensors.  The class name is wrong too.

bwspenc commented on modules/tensor_mechanics/include/materials/ADRankTwoInvariant.h:
```diff
@@ -0,0 +1,46 @@
+   */
+  MooseEnum _invariant;
+
+  /// The direction vector in which the scalar stress value is calculated
```
I don't think this needs to be a member variable. Just create a dummy Point right before the call to get the invariant.

bwspenc commented on modules/tensor_mechanics/include/materials/ADRankTwoInvariant.h:
```diff
@@ -0,0 +1,46 @@
+  ADMaterialProperty<Real> & _property;
+
+  /**
+   * Determines the information to be extracted from the tensor by using the
```
Use the `///` doxygen comment style.

bwspenc commented on modules/tensor_mechanics/include/materials/RankTwoCartesianComponent.h:
```diff
@@ -0,0 +1,43 @@
+  /// Stress/strain value returned from calculation
+  MaterialProperty<Real> & _property;
+
+  /// Tensor components
```
Doxygen grouping

bwspenc commented on modules/tensor_mechanics/include/materials/RankTwoCylindricalComponent.h:
```diff
@@ -0,0 +1,50 @@
+#include "RankTwoTensor.h"
+
+/**
+ * MaterialRankTwoCylindricalComponent uses the namespace RankTwoScalarTools to compute scalar
```
Class name and description are wrong

bwspenc commented on modules/tensor_mechanics/include/materials/RankTwoCylindricalComponent.h:
```diff
@@ -0,0 +1,50 @@
+
+  /**
+   * Determines the information to be extracted from the tensor by using the
+   * RankTwoScalarTools namespace, e.g., hoop strain, axial strain, etc.
```
Use `///` doxygen comments

bwspenc commented on modules/tensor_mechanics/include/materials/RankTwoCylindricalComponent.h:
```diff
@@ -0,0 +1,50 @@
+   */
+  MooseEnum _cylindrical_component;
+
+  /// Points used to determine the axis of rotation
```
doxygen grouping

bwspenc commented on modules/tensor_mechanics/include/materials/RankTwoCylindricalComponent.h:
```diff
@@ -0,0 +1,50 @@
+  const Point _cylindrical_axis_point2;
+
+  /// The direction vector in which the scalar stress value is calculated
+  Point _direction;
```
Don't need this

bwspenc commented on modules/tensor_mechanics/include/materials/RankTwoDirectionalComponent.h:
```diff
@@ -0,0 +1,46 @@
+   * Determines the information to be extracted from the tensor by using the
+   * RankTwoScalarTools namespace
+   */
+  MooseEnum _invariant;
```
Don't need this

bwspenc commented on modules/tensor_mechanics/include/materials/RankTwoDirectionalComponent.h:
```diff
@@ -0,0 +1,46 @@
+#include "RankTwoTensor.h"
+
+/**
+ * RankTwoDirectionalComponent uses the namespace RankTwoScalarTools to compute scalar
```
Description is wrong

bwspenc commented on modules/tensor_mechanics/include/materials/RankTwoInvariant.h:
```diff
@@ -0,0 +1,46 @@
+
+/**
+ * RankTwoInvariant uses the namespace RankTwoScalarTools to compute scalar
+ * values from Rank-2 tensors.
```
Description could be better (it's used to compute invariants)

bwspenc commented on modules/tensor_mechanics/include/materials/RankTwoInvariant.h:
```diff
@@ -0,0 +1,46 @@
+  MaterialProperty<Real> & _property;
+
+  /**
+   * Determines the information to be extracted from the tensor by using the
```
`///` doxygen

bwspenc commented on modules/tensor_mechanics/include/materials/RankTwoInvariant.h:
```diff
@@ -0,0 +1,46 @@
+  MooseEnum _invariant;
+
+  /// The direction vector in which the scalar stress value is calculated
+  Point _direction;
```
Don't need this

bwspenc commented on modules/tensor_mechanics/include/utils/RankTwoScalarTools.h:
```diff
@@ -467,4 +472,86 @@ getQuantity(const RankTwoTensorTempl<T> & tensor,
+
+template <typename T>
+T
+getCartesianComponent(const RankTwoTensorTempl<T> & tensor, const MooseEnum & scalar_type)
```
This should be something like `getInvariant`.  These aren't Cartesian components.

bwspenc commented on modules/tensor_mechanics/src/actions/TensorMechanicsAction.C:
```diff
@@ -76,7 +83,14 @@ TensorMechanicsAction::TensorMechanicsAction(const InputParameters & params)
+                                 : Point(0, 0, 0)),
+    _cylindrical_axis_point2(isParamValid("cylindrical_axis_point2")
+                                 ? getParam<Point>("cylindrical_axis_point2")
+                                 : Point(0, 1, 0)),
```
I don't think we should have defaults for this. If the user asks for hoop, radial, or axial stress, we need checks to require them to be provided.

vanwdani replied:
There are over 50 input files which are using the default points and direction.  I can change them if needed

bwspenc commented on modules/tensor_mechanics/src/actions/TensorMechanicsAction.C:
```diff
@@ -428,67 +424,131 @@ TensorMechanicsAction::actOutputGeneration()
+    {
+      std::string type = "";
+      InputParameters params = emptyInputParameters();
+      type = ad_prepend + "MaterialRealAux";
```
std::string type = ad_prepend ...
(get rid of line 434)

bwspenc commented on modules/tensor_mechanics/src/utils/RankTwoScalarTools.C:
```diff
@@ -12,6 +12,8 @@
 {
 
+///
+/// This enum is left for legacy calls
```
Delete this comment

moosebuild commented:
Job [Modules debug](https://civet.inl.gov/job/512077/) on 770c86c : invalidated by @vanwdani

bwspenc commented on modules/tensor_mechanics/doc/content/source/materials/ADRankTwoCylindricalComponent.md:
```diff
@@ -0,0 +1,28 @@
+[TensorMechanics/Master](/Modules/TensorMechanics/Master/index.md) automatically
+when stress components are requested in the generate_output parameter, but can
+also be set up directly by the user.  This class provides
+the ability to compute hoop, radial, and axial ($\boldsymbol{\sigma}$) and
```
axial stress

bwspenc commented on modules/tensor_mechanics/doc/content/source/materials/RankTwoCylindricalComponent.md:
```diff
@@ -0,0 +1,28 @@
+[TensorMechanics/Master](/Modules/TensorMechanics/Master/index.md) automatically
+when stress components are requested in the generate_output parameter, but can
+also be set up directly by the user.  This class provides the ability to compute
+hoop, radial, and axial ($\boldsymbol{\sigma}$) and strain
```
axial stress

bwspenc commented on modules/tensor_mechanics/doc/content/source/utils/RankTwoScalarTools.md:
```diff
@@ -0,0 +1,254 @@
+quantities such as invariants and components in specified directions from rank-2
+tensors such as stress or strain. These functions are not directly invoked in the
+input file, but are called by several other classes such as:
+RankTwoCylindricalComponent, RankTwoDirectionalComponent,
```
Probably good to add hyperlinks to them.

bwspenc commented on modules/tensor_mechanics/include/utils/RankTwoScalarTools.h:
```diff
@@ -467,4 +471,80 @@ getQuantity(const RankTwoTensorTempl<T> & tensor,
+
+template <typename T>
+T
+getInvariantComponent(const RankTwoTensorTempl<T> & tensor, const MooseEnum & scalar_type)
```
I'd rename this to `getInvariant()`

bwspenc reviewed (approved):
The comments are for really minor stuff. I'm going ahead and merging, but if you could get those on your next PR, that would be great. Thanks for taking care of all my comments, I'm really happy with this now!

bwspenc commented:
@dschwen I think he covered everything you wanted in your review. If you could take a quick look at this, I'd appreciate it.

dschwen commented on modules/tensor_mechanics/doc/content/source/materials/ADRankTwoCartesianComponent.md:
```diff
@@ -0,0 +1,32 @@
+# Rank Two Tensor Component
```
I'd prefer the main heading to be the class name, but that should not hold up the entire PR. I noticed that TM is going rogue with those multi word non-classname headings. Pretty much everywhere else I looked it is always teh classname... @bwspenc, any opinions?

vanwdani replied:
TM or AD?

bwspenc replied:
TM is going rogue with them (along with BISON), and I'd like to condense them down, but that's an issue for another day probably.

dschwen commented on modules/tensor_mechanics/include/materials/RankTwoCartesianComponent.h:
```diff
@@ -0,0 +1,40 @@
+#include "RankTwoTensor.h"
+
+/// ADRankTwoCartesianComponent computes selected components from a Rank-2 tensors.
+class RankTwoCartesianComponent : public Material
```
Ok, `RankTwoCartesianComponent` and `ADRankTwoCartesianComponent` should really just be typedefs of the same `RankTwoCartesianComponentTempl<bool is_ad>` class template to avoid the code duplication.

vanwdani replied:
I agree.  I did it this way so that the PR I will create tomorrow to resolve the GenericMaterialProperty can be resolved and use the typedefs

bwspenc replied:
Yes. it should. This is a temporary workaround until we can fix the issue with `getGenericMaterialProperty` giving us the wrong qp index. As soon as we merge this PR, we are planning to submit another PR that combines them using the templates, and will fail so that someone can help us fix that. I don't want to hold this capability up for that.

bwspenc commented:
Are you OK with me merging this, @dschwen?

jiangwen84 commented on modules/tensor_mechanics/src/materials/ADRankTwoInvariant.C:
```diff
@@ -0,0 +1,60 @@
+{
+  if (_property_name == "max_principal_stress" || _property_name == "mid_principal_stress" ||
+      _property_name == "min_principal_stress")
+  {
```
@bwspenc @vanwdani 

Has this to be xxx_stress? How about strain? 

The property_name is provided by user and it can be different than "max_principal_stress", correct?

vanwdani commented:
Yes.  Strain should be included.  I don't think any input file had
requested something like that previously, but I want to put it in the next
PR.  I also have additional strains/stress such as hoop_elastic_strain that
need to be added.  I didn't notice them until after I edited the input
files.

On Mon, May 4, 2020 at 5:16 PM Wen Jiang <notifications@github.com> wrote:

> *@jiangwen84* commented on this pull request.
> ------------------------------
>
> In modules/tensor_mechanics/src/materials/ADRankTwoInvariant.C
> <https://github.com/idaholab/moose/pull/14740#discussion_r419782997>:
>
> > +    _invariant(getParam<MooseEnum>("invariant"))
> +{
> +}
> +
> +void
> +ADRankTwoInvariant::initQpStatefulProperties()
> +{
> +  _property[_qp] = 0.0;
> +}
> +
> +void
> +ADRankTwoInvariant::computeQpProperties()
> +{
> +  if (_property_name == "max_principal_stress" || _property_name == "mid_principal_stress" ||
> +      _property_name == "min_principal_stress")
> +  {
>
> @bwspenc <https://github.com/bwspenc> @vanwdani
> <https://github.com/vanwdani>
>
> Has this to be xxx_stress? How about strain?
>
> The property_name is provided by user and it can be different than
> "max_principal_stress", correct?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/idaholab/moose/pull/14740#pullrequestreview-405410548>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AIAGQHGO2VC63W25OTZWFNLRP5EDZANCNFSM4KXNXNZA>
> .
>

jiangwen84 commented:
It sounds good. The hard coded "max_principal_stress", etc should be
changed.

On Mon, May 4, 2020 at 7:29 PM vanwdani <notifications@github.com> wrote:

> Yes. Strain should be included. I don't think any input file had
> requested something like that previously, but I want to put it in the next
> PR. I also have additional strains/stress such as hoop_elastic_strain that
> need to be added. I didn't notice them until after I edited the input
> files.
>
> On Mon, May 4, 2020 at 5:16 PM Wen Jiang <notifications@github.com> wrote:
>
> > *@jiangwen84* commented on this pull request.
> > ------------------------------
> >
> > In modules/tensor_mechanics/src/materials/ADRankTwoInvariant.C
> > <https://github.com/idaholab/moose/pull/14740#discussion_r419782997>:
> >
> > > + _invariant(getParam<MooseEnum>("invariant"))
> > +{
> > +}
> > +
> > +void
> > +ADRankTwoInvariant::initQpStatefulProperties()
> > +{
> > + _property[_qp] = 0.0;
> > +}
> > +
> > +void
> > +ADRankTwoInvariant::computeQpProperties()
> > +{
> > + if (_property_name == "max_principal_stress" || _property_name ==
> "mid_principal_stress" ||
> > + _property_name == "min_principal_stress")
> > + {
> >
> > @bwspenc <https://github.com/bwspenc> @vanwdani
> > <https://github.com/vanwdani>
> >
> > Has this to be xxx_stress? How about strain?
> >
> > The property_name is provided by user and it can be different than
> > "max_principal_stress", correct?
> >
> > —
> > You are receiving this because you were mentioned.
> > Reply to this email directly, view it on GitHub
> > <
> https://github.com/idaholab/moose/pull/14740#pullrequestreview-405410548>,
> > or unsubscribe
> > <
> https://github.com/notifications/unsubscribe-auth/AIAGQHGO2VC63W25OTZWFNLRP5EDZANCNFSM4KXNXNZA
> >
> > .
> >
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/idaholab/moose/pull/14740#issuecomment-623791763>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACUQYGKOCAUDTPXV77QQR7TRP5TWNANCNFSM4KXNXNZA>
> .
>

vanwdani commented:
The "hard-coded" stress actually just determines which call to the
RankTwoScalarTools will be used as seen on lines 60 ~ 67.  The principal
stresses (and in the future) strains require a dummy direction, where as
all others within the class RankTwoInvariant do not.  The boolean operation
should probably look for a string within the _property_name with something
like == *_principal_*

On Mon, May 4, 2020 at 7:56 PM Wen Jiang <notifications@github.com> wrote:

> It sounds good. The hard coded "max_principal_stress", etc should be
> changed.
>
> On Mon, May 4, 2020 at 7:29 PM vanwdani <notifications@github.com> wrote:
>
> > Yes. Strain should be included. I don't think any input file had
> > requested something like that previously, but I want to put it in the
> next
> > PR. I also have additional strains/stress such as hoop_elastic_strain
> that
> > need to be added. I didn't notice them until after I edited the input
> > files.
> >
> > On Mon, May 4, 2020 at 5:16 PM Wen Jiang <notifications@github.com>
> wrote:
> >
> > > *@jiangwen84* commented on this pull request.
> > > ------------------------------
> > >
> > > In modules/tensor_mechanics/src/materials/ADRankTwoInvariant.C
> > > <https://github.com/idaholab/moose/pull/14740#discussion_r419782997>:
> > >
> > > > + _invariant(getParam<MooseEnum>("invariant"))
> > > +{
> > > +}
> > > +
> > > +void
> > > +ADRankTwoInvariant::initQpStatefulProperties()
> > > +{
> > > + _property[_qp] = 0.0;
> > > +}
> > > +
> > > +void
> > > +ADRankTwoInvariant::computeQpProperties()
> > > +{
> > > + if (_property_name == "max_principal_stress" || _property_name ==
> > "mid_principal_stress" ||
> > > + _property_name == "min_principal_stress")
> > > + {
> > >
> > > @bwspenc <https://github.com/bwspenc> @vanwdani
> > > <https://github.com/vanwdani>
> > >
> > > Has this to be xxx_stress? How about strain?
> > >
> > > The property_name is provided by user and it can be different than
> > > "max_principal_stress", correct?
> > >
> > > —
> > > You are receiving this because you were mentioned.
> > > Reply to this email directly, view it on GitHub
> > > <
> > https://github.com/idaholab/moose/pull/14740#pullrequestreview-405410548
> >,
> > > or unsubscribe
> > > <
> >
> https://github.com/notifications/unsubscribe-auth/AIAGQHGO2VC63W25OTZWFNLRP5EDZANCNFSM4KXNXNZA
> > >
> > > .
> > >
> >
> > —
> > You are receiving this because you were mentioned.
> > Reply to this email directly, view it on GitHub
> > <https://github.com/idaholab/moose/pull/14740#issuecomment-623791763>,
> or
> > unsubscribe
> > <
> https://github.com/notifications/unsubscribe-auth/ACUQYGKOCAUDTPXV77QQR7TRP5TWNANCNFSM4KXNXNZA
> >
> > .
> >
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/idaholab/moose/pull/14740#issuecomment-623796561>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AIAGQHAQ6OEENBHGWDC7SA3RP5W55ANCNFSM4KXNXNZA>
> .
>

jiangwen84 commented:
Yes, I understand principal stress needs a dummy direction. However, the
_property_name is provided by user and we cannot determine whether it will
be something like *_principal_*. Let us discuss this more in the standup
meeting today.

On Tue, May 5, 2020 at 8:55 AM vanwdani <notifications@github.com> wrote:

> The "hard-coded" stress actually just determines which call to the
> RankTwoScalarTools will be used as seen on lines 60 ~ 67. The principal
> stresses (and in the future) strains require a dummy direction, where as
> all others within the class RankTwoInvariant do not. The boolean operation
> should probably look for a string within the _property_name with something
> like == *_principal_*
>
> On Mon, May 4, 2020 at 7:56 PM Wen Jiang <notifications@github.com> wrote:
>
> > It sounds good. The hard coded "max_principal_stress", etc should be
> > changed.
> >
> > On Mon, May 4, 2020 at 7:29 PM vanwdani <notifications@github.com>
> wrote:
> >
> > > Yes. Strain should be included. I don't think any input file had
> > > requested something like that previously, but I want to put it in the
> > next
> > > PR. I also have additional strains/stress such as hoop_elastic_strain
> > that
> > > need to be added. I didn't notice them until after I edited the input
> > > files.
> > >
> > > On Mon, May 4, 2020 at 5:16 PM Wen Jiang <notifications@github.com>
> > wrote:
> > >
> > > > *@jiangwen84* commented on this pull request.
> > > > ------------------------------
> > > >
> > > > In modules/tensor_mechanics/src/materials/ADRankTwoInvariant.C
> > > > <https://github.com/idaholab/moose/pull/14740#discussion_r419782997
> >:
> > > >
> > > > > + _invariant(getParam<MooseEnum>("invariant"))
> > > > +{
> > > > +}
> > > > +
> > > > +void
> > > > +ADRankTwoInvariant::initQpStatefulProperties()
> > > > +{
> > > > + _property[_qp] = 0.0;
> > > > +}
> > > > +
> > > > +void
> > > > +ADRankTwoInvariant::computeQpProperties()
> > > > +{
> > > > + if (_property_name == "max_principal_stress" || _property_name ==
> > > "mid_principal_stress" ||
> > > > + _property_name == "min_principal_stress")
> > > > + {
> > > >
> > > > @bwspenc <https://github.com/bwspenc> @vanwdani
> > > > <https://github.com/vanwdani>
> > > >
> > > > Has this to be xxx_stress? How about strain?
> > > >
> > > > The property_name is provided by user and it can be different than
> > > > "max_principal_stress", correct?
> > > >
> > > > —
> > > > You are receiving this because you were mentioned.
> > > > Reply to this email directly, view it on GitHub
> > > > <
> > >
> https://github.com/idaholab/moose/pull/14740#pullrequestreview-405410548
> > >,
> > > > or unsubscribe
> > > > <
> > >
> >
> https://github.com/notifications/unsubscribe-auth/AIAGQHGO2VC63W25OTZWFNLRP5EDZANCNFSM4KXNXNZA
> > > >
> > > > .
> > > >
> > >
> > > —
> > > You are receiving this because you were mentioned.
> > > Reply to this email directly, view it on GitHub
> > > <https://github.com/idaholab/moose/pull/14740#issuecomment-623791763>,
> > or
> > > unsubscribe
> > > <
> >
> https://github.com/notifications/unsubscribe-auth/ACUQYGKOCAUDTPXV77QQR7TRP5TWNANCNFSM4KXNXNZA
> > >
> > > .
> > >
> >
> > —
> > You are receiving this because you were mentioned.
> > Reply to this email directly, view it on GitHub
> > <https://github.com/idaholab/moose/pull/14740#issuecomment-623796561>,
> or
> > unsubscribe
> > <
> https://github.com/notifications/unsubscribe-auth/AIAGQHAQ6OEENBHGWDC7SA3RP5W55ANCNFSM4KXNXNZA
> >
> > .
> >
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/idaholab/moose/pull/14740#issuecomment-624104542>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACUQYGIMROIXRZE4LHOGKPTRQASFXANCNFSM4KXNXNZA>
> .
>
